File: algo/src/main/java/com/hartwig/actin/algo/evaluation/general/HasMinimumLanskyKarnofskyScore.java
package com.hartwig.actin.algo.evaluation.general;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;

import org.jetbrains.annotations.NotNull;

public class HasMinimumLanskyKarnofskyScore implements EvaluationFunction {

    @NotNull
    private final PerformanceScore performanceScore;
    private final int minScore;

    HasMinimumLanskyKarnofskyScore(@NotNull final PerformanceScore performanceScore, final int minScore) {
        this.performanceScore = performanceScore;
        this.minScore = minScore;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Integer who = record.clinical().clinicalStatus().who();
        if (who == null) {
            return EvaluationFactory.recoverable()
                    .result(EvaluationResult.UNDETERMINED)
                    .addUndeterminedSpecificMessages(
                            "Cannot evaluate " + performanceScore.display() + " performance score because WHO is missing")
                    .addUndeterminedGeneralMessages("Missing " + performanceScore.display() + " score")
                    .build();
        }

        int passScore = toMinScoreForWHO(who);
        int undeterminedScore = toMaxScoreForWHO(who);
        int warnScore = toMaxScoreForWHO(Math.max(0, who - 1));

        if (passScore >= minScore) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addPassSpecificMessages(performanceScore.display() + " score based on WHO score is at least " + minScore)
                    .addPassGeneralMessages("Minimum " + performanceScore.display() + " requirements")
                    .build();
        } else if (undeterminedScore >= minScore) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.UNDETERMINED)
                    .addUndeterminedSpecificMessages(
                            "Not clear whether " + performanceScore.display() + " score based on WHO score is at least " + minScore)
                    .addUndeterminedSpecificMessages("Undetermined minimum " + performanceScore.display() + " requirements")
                    .build();
        } else if (warnScore >= minScore) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.WARN)
                    .addWarnSpecificMessages(performanceScore.display() + " score based on WHO score exceeds requested score of " + minScore)
                    .addWarnSpecificMessages("Minimum " + performanceScore.display() + " requirements")
                    .build();
        }

        return EvaluationFactory.unrecoverable()
                .result(EvaluationResult.FAIL)
                .addFailSpecificMessages(performanceScore.display() + " score based on WHO score is below " + minScore)
                .addFailGeneralMessages("Minimum " + performanceScore.display() + " requirements")
                .build();
    }

    private static int toMinScoreForWHO(int who) {
        switch (who) {
            case 0:
                return 100;
            case 1:
                return 80;
            case 2:
                return 60;
            case 3:
                return 40;
            case 4:
                return 10;
            case 5:
                return 0;
            default:
                throw new IllegalStateException("Illegal who status: " + who);
        }
    }

    private static int toMaxScoreForWHO(int who) {
        switch (who) {
            case 0:
                return 100;
            case 1:
                return 90;
            case 2:
                return 70;
            case 3:
                return 50;
            case 4:
                return 30;
            case 5:
                return 0;
            default:
                throw new IllegalStateException("Illegal who status: " + who);
        }
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/general/HasMinimumLanskyKarnofskyScore.kt
package com.hartwig.actin.algo.evaluation.general

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.clinical.interpretation.asRange
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation

class HasMinimumLanskyKarnofskyScore internal constructor(
    private val performanceScore: PerformanceScore,
    private val minScore: Int,
    private val labels: EvaluationLabels.General
) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val scoreDisplay = performanceScore.display()
        val who = record.performanceStatus.latestWho
            ?: return EvaluationFactory.undetermined(labels.hasMinimumLanskyKarnofskyScoreUndeterminedMissing(scoreDisplay, minScore))
        val whoRange = who.asRange()
        val passScore = toMinScoreForWHO(whoRange.last)
        val undeterminedScore = toMaxScoreForWHO(whoRange.first)
        val warnScore = toMaxScoreForWHO((whoRange.first - 1).coerceAtLeast(0))

        return when {
            passScore >= minScore -> {
                EvaluationFactory.pass(labels.hasMinimumLanskyKarnofskyScorePass(scoreDisplay, minScore))
            }

            undeterminedScore >= minScore -> {
                EvaluationFactory.undetermined(labels.hasMinimumLanskyKarnofskyScoreUndetermined(scoreDisplay, minScore))
            }

            warnScore >= minScore -> {
                EvaluationFactory.recoverableFail(labels.hasMinimumLanskyKarnofskyScoreRecoverableFail(scoreDisplay, minScore))
            }

            else -> EvaluationFactory.fail(labels.hasMinimumLanskyKarnofskyScoreFail(scoreDisplay, minScore))
        }
    }

    companion object {
        private fun toMinScoreForWHO(who: Int): Int {
            return when (who) {
                0 -> 100
                1 -> 80
                2 -> 60
                3 -> 40
                4 -> 10
                5 -> 0
                else -> throw IllegalStateException("Illegal who status: $who")
            }
        }

        private fun toMaxScoreForWHO(who: Int): Int {
            return when (who) {
                0 -> 100
                1 -> 90
                2 -> 70
                3 -> 50
                4 -> 30
                5 -> 0
                else -> throw IllegalStateException("Illegal who status: $who")
            }
        }
    }
}